### PR TITLE
Allow for fetching packages from remote git repositories

### PR DESCRIPTION
**Pull request by nabuskey**
This allows for fetching packages from remote git repositories.

The URL format follows kustomize remote url format.

fixes: #184 

example:
```
idpbuilder create  --package-dir https://github.com/cnoe-io/idpbuilder//examples/basic/package1?version=v0.2.0
```

This should also reduce the number of git clone operations against local and remote repositories.

### CODE REVIEW

File: pkg/controllers/gitrepository/gitea.go [hunk 1]
```diff
@@ -97,8 +98,15 @@ func (g *giteaProvider) getRepository(ctx context.Context, repo *v1alpha1.GitRep
 	}, nil
 }
 
-func (g *giteaProvider) updateRepoContent(ctx context.Context, repo *v1alpha1.GitRepository, repoInfo repoInfo, creds gitProviderCredentials) error {
-	return updateRepoContent(ctx, repo, repoInfo, creds, g.Scheme, g.config)
+func (g *giteaProvider) updateRepoContent(ctx context.Context, repo *v1alpha1.GitRepository, repoInfo repoInfo, creds gitProviderCredentials, tmpDir string, repoMap *util.RepoMap) error {
```
Comment (nimakaviani) [inline pkg/controllers/gitrepository/gitea.go:101]:
```suggestion
func (g *giteaProvider) updateRepoContent(
    ctx context.Context,
    repo *v1alpha1.GitRepository,
    repoInfo repoInfo,
    creds gitProviderCredentials,
    tmpDir string,
    repoMap *util.RepoMap,
) error {
```

File: pkg/controllers/gitrepository/github.go [hunk 2]
```diff
@@ -102,8 +102,8 @@ func (g *gitHubProvider) setProviderCredentials(ctx context.Context, repo *v1alp
 	return g.gitHubClient.setToken(creds.accessToken)
 }
 
-func (g *gitHubProvider) updateRepoContent(ctx context.Context, repo *v1alpha1.GitRepository, repoInfo repoInfo, creds gitProviderCredentials) error {
-	return updateRepoContent(ctx, repo, repoInfo, creds, g.Scheme, g.config)
+func (g *gitHubProvider) updateRepoContent(ctx context.Context, repo *v1alpha1.GitRepository, repoInfo repoInfo, creds gitProviderCredentials, tmpDir string, repoMap *util.RepoMap) error {
```
Comment (nimakaviani) [inline pkg/controllers/gitrepository/github.go:105]:
```suggestion
func (g *gitHubProvider) updateRepoContent(
    ctx context.Context,
    repo *v1alpha1.GitRepository,
    repoInfo repoInfo,
    creds gitProviderCredentials,
    tmpDir string,
    repoMap *util.RepoMap,
) error {
```

File: pkg/controllers/localbuild/controller.go [hunk 3]
```diff
@@ -140,12 +141,18 @@ func (r *LocalbuildReconciler) ReconcileArgoAppsWithGitea(ctx context.Context, r
...
 	}
-	if resource.Spec.PackageConfigs.CustomPackageDirs != nil {
-		for i := range resource.Spec.PackageConfigs.CustomPackageDirs {
-			result, err := r.reconcileCustomPkg(ctx, resource, resource.Spec.PackageConfigs.CustomPackageDirs[i])
-			if err != nil {
-				return result, err
-			}
+
+	for i := range resource.Spec.PackageConfigs.CustomPackageDirs {
+		result, err := r.reconcileCustomPkgDir(ctx, resource, resource.Spec.PackageConfigs.CustomPackageDirs[i])
```
Comment (nimakaviani) [inline pkg/controllers/localbuild/controller.go:146]:
```suggestion
	for _, d := range resource.Spec.PackageConfigs.CustomPackageDirs {
		result, err := r.reconcileCustomPkgDir(ctx, resource, d)
```

File: pkg/controllers/localbuild/controller.go [hunk 4]
```diff
@@ -284,7 +290,105 @@ func (r *LocalbuildReconciler) shouldShutDown(ctx context.Context, resource *v1a
...
+		return ctrl.Result{}, fmt.Errorf("cloning repo, %s: %w", pkgUrl, err)
+	}
+
+	yamlFiles, err := util.GetWorktreeYamlFiles(remote.Path(), wt, false)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting yaml files from repo, %s: %w", pkgUrl, err)
+	}
+
+	for i := range yamlFiles {
+		n := yamlFiles[i]
```
Comment (nimakaviani) [inline pkg/controllers/localbuild/controller.go:322]:
```suggestion
	for _, n := range yamlFiles {
```

File: pkg/controllers/run.go [hunk 5]
```diff
@@ -12,16 +12,20 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
-func RunControllers(ctx context.Context, mgr manager.Manager, exitCh chan error, ctxCancel context.CancelFunc, exitOnSync bool, cfg util.CorePackageTemplateConfig) error {
+func RunControllers(ctx context.Context, mgr manager.Manager, exitCh chan error, ctxCancel context.CancelFunc, exitOnSync bool, cfg util.CorePackageTemplateConfig, tmpDir string) error {
```
Comment (nimakaviani) [inline pkg/controllers/run.go:15]:
```suggestion
func RunControllers(
    ctx context.Context,
    mgr manager.Manager,
    exitCh chan error,
    ctxCancel context.CancelFunc,
    exitOnSync bool,
    cfg util.CorePackageTemplateConfig,
    tmpDir string,
) error {
```

File: pkg/util/git_repository.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+func CopyTreeToTree(srcWT, dstWT billy.Filesystem, srcPath, dstPath string) error {
+	files, err := srcWT.ReadDir(srcPath)
+	if err != nil {
+		return err
+	}
+
+	for i := range files {
+		srcFile := files[i]
```
Comment (nimakaviani) [inline pkg/util/git_repository.go:192]:
```suggestion
	for _, srcFile := range files {
```

**Review by nimakaviani — COMMENTED** [review]
Thanks for this left a few comments.

I think we should also update the README to go with it. Can you please make the necessary changes?

File: pkg/controllers/gitrepository/controller.go [hunk 7]
```diff
@@ -189,94 +191,162 @@ func (r *RepositoryReconciler) SetupWithManager(mgr ctrl.Manager, notifyChan cha
...
+			return fmt.Errorf("pushing to git: %w", err)
+		}
+
+		repo.Status.LatestCommit.Hash = hash.String()
+		return nil
 	}
 
-	repo.Status.LatestCommit.Hash = commit.String()
+	repo.Status.LatestCommit.Hash = hash.String()
 	return nil
```
[the remark this answers is not in the document]
Comment (nabuskey) [reply pkg/controllers/gitrepository/controller.go:350]: It does looks similar but I don't think introducing another function would save us many lines. Maybe like 3 lines. I think it's pretty clear what these functions are doing right now and I'd rather keep it that way.

File: pkg/controllers/custompackage/controller.go [hunk 8]
```diff
@@ -91,17 +92,17 @@ func (r *Reconciler) reconcileCustomPackage(ctx context.Context, resource *v1alp
 
 	app, ok := objs[0].(*argov1alpha1.Application)
 	if !ok {
-		return ctrl.Result{}, fmt.Errorf("object is not an PackageSpec application %s", resource.Spec.ArgoCD.ApplicationFile)
+		return ctrl.Result{}, fmt.Errorf("object is not an ArgoCD application %s", resource.Spec.ArgoCD.ApplicationFile)
```
[the remark this answers is not in the document]
Comment (nabuskey) [reply pkg/controllers/custompackage/controller.go:95]: I think it used to be ArgoCD but I must have let the auto refactor change it in the past. The error message should say it's not an ArgoCD application.

File: README.md [hunk 9]
```diff
@@ -153,6 +156,12 @@ To deploy these packages, run the following commands from this repository's root
 ./idpbuilder create --package-dir examples/basic/package1  --package-dir examples/basic/package2
 ```
 
+Alternatively, you can use the URL format:
+
+```
+./idpbuilder create  --package-dir https://github.com/cnoe-io/idpbuilder//examples/basic/package1 --package-dir https://github.com/cnoe-io/idpbuilder//examples/basic/package2
```
[the remark this answers is not in the document]
Comment (nimakaviani) [reply README.md:162]: we probably should just make it `--package` and it can take a dir or a url

[the remark this answers is not in the document]
Comment (nabuskey) [reply README.md:162]: https://github.com/cnoe-io/idpbuilder/issues/264 Opened for this.

**Review by nimakaviani — APPROVED** [review]
given that we have #264 to add a new flag to handle package references, I am going to approve and merge this, so we can unblock the rest of the work.

**Pull request merged**